### ISSUE TITLE
NUnit hangs when reporting to TeamCity

### ISSUE THREAD

**Issue by martinskuta** [nunit/nunit#1432]
Recently we have updated our build to be run by Cake (C# make) script which runs our tests through dotCover from JetBrains on TeamCity agent. We have noticed that NUnit automagically realizes it is run on team city agent and reports to the server, which is cool, but it seems that it can make the nunit-console-runner to hang. I think it is related to to that, because after inspecting the process dump there are two worker threads doing Socket.Recieve and other two doing TeamCityServiceMessagePublisher.RegisterMessage. 

Some more facts:

- The hang is very random, 90% of the time our build finishes fine. 
- When it is stuck it is stuck untill it is killed (24h)
- NUnit 3.2.1, ran via dotCover on WindowsServer 2012R2 x64, CLR 4.6.1055
- we have four dlls containing tests, these are run in parallel

![process explorer](https://dl.dropboxusercontent.com/u/6537800/nunit/process_explorer_scr_when_stuck.png "process explorer screenshot")

[full process dump](https://dl.dropboxusercontent.com/u/6537800/nunit/nunit3-console.dmp)

Let me know if you need more info.

**Comment by rprouse** [comment]
Are you using any of the parallel features in NUnit other than parallel assemblies?

**Comment by martinskuta** [comment]
> Are you using any of the parallel features in NUnit other than parallel assemblies?
No, just parallel assemblies. Maybe what might be relevant is that we have some async tests `public async Task` and some that requires STA, but they run sequentially.

**Comment by martinskuta** [comment]
Today we experienced another build stuck while running tests. 75% of CPU (3 of 4 cpu cores) have been full at work! Looking at the process dump it is same issue again.. 
![image](https://cloud.githubusercontent.com/assets/14070352/14743021/9390e84a-089f-11e6-8448-56667829a2b4.png)

Yesterday it was 50% CPU and two threads doing Socket.Recieve and one doing TeamCityServiceMessagePublisher.RegisterMessage and one doing TryFindPartnerId. You can see on the screenshot today it were three threads doing Socket.Recieve, two doing TeamCityServiceMessagePublisher.RegisterMessage and one doing TryPartnerId ... so RegisterMessage and TryFindPartnerId there seems to lie the issue. Seems like the RegisterMessage and TryFindPartnerId are deadlocking ?

It is kind of annoying since we cannot find the root cause. I will start looking at NUnit code when won't be in release week. Someone some idea?

**Comment by martinskuta** [comment]
We will try to generate the reports and import it to TC ourselves. Lets see if it solves the issue.

**Comment by rprouse** [comment]
As a workaround, I expect that running the tests and importing the reports yourself will work. The JetBrains developers contributed the TeamCity specific code to NUnit, so we are not as familiar with it as we should be. To be honest, we were a bit reluctant to accept CI specific code into NUnit since no other CI system requires it. A bit of a slippery slope :wink:

**Comment by laurentkempe** [comment]
@rprouse I had the same reaction when I read that it would magically detect TeamCity and not other build servers!

**Comment by martinskuta** [comment]
Ok so even if we don't pass the --teamcity and import the test results ourselves nunit still gets stuck:
![image](https://cloud.githubusercontent.com/assets/14070352/14916318/4f2e4632-0e19-11e6-98e8-496be4cb3723.png)

Same issue :(

**Comment by martinskuta** [comment]
I just found out that the team city integration cannot be turned off... because if I pass --teamcity or not to the console runner it doesn't change anything, it is still on, because of the `NUnit.Common.DefaultOptionsProvider` which defaults to true on our team city agent.

**Comment by rprouse** [comment]
You've done enough work on this that I am going to move into the backlog. Hopefully one of the TeamCity developers who originally added this feature/code can take a look? Of course, anyone else is free to take it and run with it :racehorse: :smile:

**Comment by CharliePoole** [comment]
@rprouse When we accepted this feature, I made the assumption that the setting of the environment variable was under the user's control.  If it really isn't, then I think we should provide a way to defeat it.

**Comment by rprouse** [comment]
@CharliePoole I agree. We have another issue to provide a way to defeat it, #1469

**Comment by NikolayPianikov** [comment]
As I wrote in #1469:
To turn off the integration with TeamCity you could add the environment variable "env.TEAMCITY_PROJECT_NAME" with empty value to your build configuration or build project.

I am going to look at this issue.
@CharliePoole @rprouse  please assign similar problems to me.

**Comment by NikolayPianikov** [comment]
TeamCityServiceMessagePublisher does not provide thread safety, so it does not use any sync primitives. 

Also it supposes that an instance of the class "TestEventHandler" have to invoke "OnTestEvent" successively and the some invoker provides thread safety. Do threads 21708 and 9024 (see screenshot) use different instances of TeamCityServiceMessagePublisher? @CharliePoole is it correct assumption? I can check it, but may be it is obvious.

@martinskuta deadlocks are not a reason of consuming CPUs.
TeamCityServiceMessagePublisher uses System.Collections.Generic.Dictionary so it have to work quite fast. Could you attach the xml report of your tests for the further investigation. Could you run tests with the arg "--teamcity" and redirect all output to file. Please attach this file too.

PS: TeamCity has ability to show feedback in real-time for tests. I am not sure that other build servers has the same ability. That's why this integration is very important.

**Comment by NikolayPianikov** [comment]
@martinskuta could you try to run without dotCover. Has anything changed?
Also try to use additional args "--workers=1 --agents=1"

**Comment by martinskuta** [comment]
> @martinskuta deadlocks are not a reason of consuming CPUs.

@NikolayPianikov yeah clear, cannot be deadlock. That was my guess at first without thinking too much, but you are right. You know it is quite hard to say, that certain things solved the issue, because this hang is very random. For now I will disable the automagical message publishing, with your provided workaround, as we dont look at the "live" progress of tests anyway, we just want to see pass/fail and if fail, break the build and see what test failed. I will come back if the hang will happen again.

**Comment by rprouse** [comment]
Thanks for stepping in @NikolayPianikov. I have assigned this to you. Let me know if you need help.

**Comment by NikolayPianikov** [comment]
As workaround you could add arg "--agents=1""
